File: hospital-management-system/backend/src/main/java/com/dhananjay/hospitalmanagement/service/DoctorService.java
package com.dhananjay.hospitalmanagement.service;

import java.util.List;
import java.util.NoSuchElementException;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import com.dhananjay.hospitalmanagement.enums.AppointmentStatus;
import com.dhananjay.hospitalmanagement.exceptions.AppointmentNotFoundException;
import com.dhananjay.hospitalmanagement.model.Appointment;
import com.dhananjay.hospitalmanagement.model.Doctor;
import com.dhananjay.hospitalmanagement.repository.AppoinmentRepository;
import com.dhananjay.hospitalmanagement.repository.DoctorRepository;
import com.dhananjay.hospitalmanagement.security.Users;

@Service
public class DoctorService {
	DoctorRepository doctorRepository;
	AppoinmentRepository appointmentRepository;
	UsersService userService;
	
 	public DoctorService(DoctorRepository doctorRepository, AppoinmentRepository appointmentRepository,UsersService userService) {
 		this.doctorRepository = doctorRepository;
		this.appointmentRepository = appointmentRepository;
		this.userService = userService;
	}

	//Add Doctor
	public Doctor addDoctor (Doctor doctor) {
		Users user = userService.registerDoctor(doctor.getUser());
		doctor.setUser(user);
		doctor.getUser().setEmail(doctor.getEmail());
		return doctorRepository.save(doctor);
 	}
	
	public Doctor getDoctorByUsername(String username) {

	    return doctorRepository.findByUser_Username(username)
	            .orElseThrow(() ->
	                new NoSuchElementException(
	                    "Doctor not found for username " + username
	                )
	            );
	}
	//Add multiple doctor at once
	public List<Doctor> addMultipleDoctor(List<Doctor> doctor){
		return doctorRepository.saveAll(doctor);
	}
	
	//Get All Doctors
	public List<Doctor> getAllDoctors (){
		return doctorRepository.findAll();
	}
	
	//Find by Id
	public Doctor findById (Long id) {
  		    Doctor doctor = doctorRepository.findById(id).orElse(null);

		    if (doctor == null) {
		        throw new NoSuchElementException("Doctor not found with id: " + id);
		    }
 		    return doctor;
	}
	
	//Delete Doctor
	public void deleteDoctor (Long id) {
		Doctor doctor  = findById(id);
		
		if(doctor.getAppointments().isEmpty()) {
			doctorRepository.delete(doctor);
		}
		else {
			throw new ResponseStatusException(
				    HttpStatus.CONFLICT,
				    "Cannot delete doctor because appointments exist"
				);
		}
	}
	
	//Update Doctor
	public Doctor updateDoctor (Doctor updatedDoctor,Long id) {
		Doctor doctor = findById(id);
		doctor.setFirstName(updatedDoctor.getFirstName());
		doctor.setLastName(updatedDoctor.getLastName());
		doctor.setSpecialization(updatedDoctor.getSpecialization());
		doctor.setEmail(updatedDoctor.getEmail());
		doctor.setPhoneNumber(updatedDoctor.getPhoneNumber());
		doctor.setQualification(updatedDoctor.getQualification());
		doctor.setExperienceYears(updatedDoctor.getExperienceYears());
		Users user = updatedDoctor.getUser();
		Users updatedUser = userService.updateDoctorUser(user,user.getId());
		doctor.setUser(updatedUser);
		
		return doctorRepository.save(doctor);
 	}
	
	//GetAppointments By Doctor Id
	public List<Appointment> getAppointmentByDoctorId(Long id) {
		 List<Appointment> appointments= appointmentRepository.findByDoctorId(id);
		 return appointments;
 	}
	
	//UpdateAppointment Status
 	public Appointment updateAppointmentStatus (Long doctorId,Long appointmentId,AppointmentStatus status) {
		 
 		Appointment appointment = appointmentRepository.findByIdAndDoctorId(appointmentId,doctorId);
 		if(appointment == null)
 			throw new AppointmentNotFoundException("Appointment Not Found");
 		else
 			appointment.setAppointmentStatus(status);
 	
 		appointmentRepository.save(appointment);
 		return appointment;
 	}
	
	
}

File: hospital-management-system/frontend/src/app/service/doctorService/doctor-service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { AppointmentEntity } from '../../model/appointment/appointment';
import { DoctorEntity } from '../../model/doctor/doctor';

@Injectable({
  providedIn: 'root',
})
export class DoctorService {
  url = 'http://localhost:8080/api/v1/doctors';

  constructor(private httpClient: HttpClient) {}

  getAllDoctors() {
    return this.httpClient.get<DoctorEntity[]>(`${this.url}`);
  }

  getDoctorById(id: any) {
    return this.httpClient.get<DoctorEntity>(`${this.url}/${id}`);
  }

  getAppointmentByDoctorId(id: any) {
    return this.httpClient.get<AppointmentEntity[]>(`${this.url}/appointments/${id}`);
  }

  addDoctor(doctor: DoctorEntity) {
    return this.httpClient.post<DoctorEntity>(`${this.url}/register`, doctor);
  }

  updateDoctor(doctor: DoctorEntity, id: any) {
    return this.httpClient.put<DoctorEntity>(`${this.url}/${id}`, doctor);
  }
  getLoggedInDoctor() {
    return this.httpClient.get<DoctorEntity>(`${this.url}/me`);
  }
  deleteDoctor(id: any) {
    return this.httpClient.delete<string>(`${this.url}/${id}`);
  }
  updateAppointmentStatus(doctorId: number, appointmentId: number, status: string) {
    return this.httpClient.put<AppointmentEntity>(
      `${this.url}/${doctorId}/appointments/${appointmentId}/${status}`,
      {},
    );
  }
  
}
